WordPress/gutenberg
# ButtonGroup: Convert to TypeScript

## What?
Converts the `ButtonGroup` component to TypeScript.

## Why?
Part of the @wordpress/components's TypeScript migration (#35744).

## How?
Refactors the current `ButtonGroup` to TypeScript.

## Testing Instructions
- Run npm run dev and ensure no typing errors
- Fire up the Storybook and confirm the `ButtonGroup` continues to function as expected
- Test the various button groups within the block/site editors in Gutenberg still work
